py/server/deephaven/jcompat.py

```diff
@@ -255,26 +255,51 @@ def _j_array_to_numpy_array(dtype: DType, j_array: jpy.JType, conv_null: bool, t
         np_array = np.array(j_array, np.object_)
 
     if conv_null:
-        if dh_null := _PRIMITIVE_DTYPE_NULL_MAP.get(dtype):
-            if dtype in (dtypes.float32, dtypes.float64):
-                np_array = np.copy(np_array)
-                np_array[np_array == dh_null] = np.nan
-            else:
-                if dtype is dtypes.bool_:  # needs to change its type to byte for dh null detection
-                    np_array = np.frombuffer(np_array, np.byte)
-
-                if any(np_array[np_array == dh_null]):
-                    if not type_promotion:
-                        raise DHError(f"Problem creating numpy array.  Java {dtype} array contains Deephaven null values, but numpy {np_array.dtype} array does not support null values")
-                    np_array = np_array.astype(np.float64)
-                    np_array[np_array == dh_null] = np.nan
-                else:
-                    if dtype is dtypes.bool_:  # needs to change its type back to bool
-                        np_array = np.frombuffer(np_array, np.bool_)
-                    return np_array
+        return dh_null_to_nan(np_array, type_promotion)
 
     return np_array
 
+def dh_null_to_nan(np_array: np.ndarray, type_promotion: bool = True) -> np.ndarray:
```

Comment:
1. In the above function, `type_promotion` defaults to false, but here it defaults to true.  Why the difference?
2. `type_promotion` defaulting to true is defaulting to the least performant and possibly most surprising default.  What is the justification for one default over another?
3. We should also be providing an in-place argument or function for better performance.

Reply:
1 and 2,  Default to False in _j_array_to_numpy_array is for performance. But If a user has to explicitly call this function, we are probably safe to assume that he/she knows 'nan' is a floating type concept and wouldn't be surprised to see the returned array is a floating type even if the input is of an integer type. In addition, the name of the function implies the conversion/promotion as well.
3. documented the in-place behavior of the function to avoid surprises, don't really think we need another optional param as we are already do the conversion the most performant way possible.